crates/sui-indexer/src/handlers/pruner.rs

```diff
@@ -112,3 +106,167 @@ impl Pruner {
         Ok(())
     }
 }
+
+/// Check if the lowest epoch of unpruned data exceeds the retention policy.
+fn should_update_watermark(watermark: &WatermarkRead, epochs_to_keep: u64) -> bool {
+    watermark.epoch_lo + epochs_to_keep <= watermark.epoch_hi
+}
+
+/// Determine if pruning should occur based on the current watermark state. When an entity's `hi`
+/// and `lo` (`reader_lo`) watermarks are updated to reflect the latest data range, if the
+/// `reader_lo` value is at least 1 greater than `pruned_lo`, then the pruner should prune. When
+/// `lo` is updated, and `pruned_lo` is still unset, that also means the pruner should start
+/// pruning.
+fn should_prune(watermark: &WatermarkRead) -> bool {
+    match watermark.pruned_lo {
+        None => watermark.reader_lo() > 0,
+        Some(pruned_lo) => watermark.reader_lo() > pruned_lo + 1,
+    }
+}
+
+/// Pruner waits for some time before pruning to ensure that in-flight reads complete or timeout
+/// before the underlying data is pruned.
+async fn wait_for_prune_delay(
+    watermark: &WatermarkRead,
+    cancel: &CancellationToken,
+    delay_amount: i64,
+) -> IndexerResult<()> {
+    let current_time = chrono::Utc::now().timestamp_millis();
```

Comment:
Time comes from the DB, always -- read it in the same query that reads the watermark data, and add a function to `WatermarkRead` that will tell you whether a given `Duration` has passed.